kore/test/Test/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,130 @@
+import Kore.Internal.TermLike hiding ( substitute )
+import Kore.Rewriting.RewritingVariable
+import Kore.Rewriting.UnifyingRule
+import Kore.Step.Implication
+
+import qualified Test.Kore.Step.MockSymbols as Mock
+
+test_freeVariables :: TestTree
+test_freeVariables =
+    testCase "Extract free variables" $ do
+        let expect =
+                foldMap (freeVariable . mkSomeVariable)
+                [x, z, t]
+            actual = freeVariables testRulePattern
+        assertEqual "Expected free variables" expect actual
+
+test_refreshRule :: [TestTree]
+test_refreshRule =
+    [ testCase "Rename target variables" $ do
+        let avoiding :: FreeVariables RewritingVariableName
+            avoiding = freeVariables testRulePattern
+            (renaming, rulePattern') =
+                refreshRule avoiding testRulePattern
+            renamed = Set.fromList (Foldable.toList renaming)
+            free' :: FreeVariables RewritingVariableName
+            free' = freeVariables rulePattern'
+            notAvoided (variableName -> var) =
+                not (FreeVariables.isFreeVariable var avoiding)
+        assertEqual
+            "Expected to rename all free variables of original RulePattern"
+            (FreeVariables.toNames avoiding)
+            (Map.keysSet renaming)
+        assertBool
+            "Expected to renamed variables distinct from original variables"
+            (all notAvoided renamed)
+        assertBool
+            "Expected no free variables in common with original RulePattern"
+            (all notAvoided (FreeVariables.toList free'))
+    , testCase "no stale variables" $ do
+        let (renaming, _) = refreshRule mempty testRulePattern
+        assertBool "expected not to rename variables" (null renaming)
+    , testGroup "stale existentials" $
+        let assertions (renaming, claim@Implication { existentials }) = do
+                assertBool "expected to refresh existentials"
+                    (notElem y existentials)
+                assertBool "expected to substitute fresh variables"
+                    (notElem (inject y) $ FreeVariables.toList (freeVariablesRight claim))
+                assertBool "expected not to rename free variables"
+                    (null renaming)
+        in
+        [ testCase "from outside" $ do
+            let stale = freeVariable (inject y)
+            assertions $ refreshRule stale testRulePattern
+        , testCase "from left-hand side" $ do
+            let input =
+                    testRulePattern
+                        { left =
+                            Pattern.fromTermAndPredicate
+                                (mkElemVar y)
+                                ( Predicate.makeCeilPredicate_
+                                    (mkElemVar z)
+                                )
+                        }
+            assertions $ refreshRule mempty input
+        ]
+    ]
+
+test_substituteRightIsntCapturingFreeVars :: TestTree
```

Comment:
I will only be updating this file as the old one has to go anyway.
Also note that I updated this test and it is different from the one in ClaimPattern (for the same reason as above), just in case you didn't notice. To that end, can you please confirm that you're okay with this new test (particularly the assertion at the end)?